File: frontend/src/clusterMerge.ts
// Client-side port of backend/app/cluster_merge.py.
//
// Used by the Upload tab so the merge sliders give instant feedback
// instead of round-tripping to /api/predict (which re-runs the model
// each call, ~100 ms+). The backend exposes /api/predict?merge=false
// to skip its own pass; the result here is identical to what the
// backend would have produced for the same parameters.
//
// Keep this in sync with the Python implementation. Behavior tests
// live next to that file.

import type { Detection } from './api'

export type MergeParams = {
  enabled: boolean
  marginFrac: number    // 0..~0.05, fraction of max(imageW, imageH)
  minSize: number       // smallest component that becomes a cluster (≥2)
  sourceClassName: string
  targetClassName: string
}

type BBox = [number, number, number, number]

function inflate([x1, y1, x2, y2]: BBox, m: number): BBox {
  return [x1 - m, y1 - m, x2 + m, y2 + m]
}

function intersects(a: BBox, b: BBox): boolean {
  return !(a[2] < b[0] || b[2] < a[0] || a[3] < b[1] || b[3] < a[1])
}

class UnionFind {
  private parent: number[]
  constructor(n: number) { this.parent = Array.from({ length: n }, (_, i) => i) }
  find(i: number): number {
    let r = i
    while (this.parent[r] !== r) r = this.parent[r]
    while (this.parent[i] !== r) { const next = this.parent[i]; this.parent[i] = r; i = next }
    return r
  }
  union(i: number, j: number): void {
    const ri = this.find(i), rj = this.find(j)
    if (ri !== rj) this.parent[ri] = rj
  }
}

export function mergeClusters(
  detections: Detection[],
  imageSize: [number, number],
  params: MergeParams,
): Detection[] {
  if (!params.enabled || params.minSize < 2 || detections.length === 0) {
    return detections.slice()
  }

  const src = detections.filter(d => d.class_name === params.sourceClassName)
  const other = detections.filter(d => d.class_name !== params.sourceClassName)
  if (src.length < params.minSize) return detections.slice()

  const [imgW, imgH] = imageSize
  const marginPx = params.marginFrac * Math.max(imgW, imgH)
  const inflated = src.map(d => inflate(d.bbox as BBox, marginPx))

  const uf = new UnionFind(src.length)
  for (let i = 0; i < src.length; i++) {
    const ai = inflated[i]
    for (let j = i + 1; j < src.length; j++) {
      if (intersects(ai, inflated[j])) uf.union(i, j)
    }
  }

  // Group source-class boxes by component root.
  const groups = new Map<number, number[]>()
  for (let i = 0; i < src.length; i++) {
    const r = uf.find(i)
    const arr = groups.get(r); if (arr) arr.push(i); else groups.set(r, [i])
  }

  // Use the first non-source class_id from the model's output we've
  // ever seen, if any, so the cluster colour stays distinct. Otherwise
  // synthesize one past the highest class_id in this response.
  let targetClassId = src[0].class_id + 1
  for (const d of detections) {
    if (d.class_id > targetClassId) targetClassId = d.class_id + 1
  }

  const out: Detection[] = []
  for (const members of groups.values()) {
    if (members.length >= params.minSize) {
      let x1 = Infinity, y1 = Infinity, x2 = -Infinity, y2 = -Infinity, c = 0
      for (const i of members) {
        const [a, b, cR, dR] = src[i].bbox
        if (a < x1) x1 = a; if (b < y1) y1 = b
        if (cR > x2) x2 = cR; if (dR > y2) y2 = dR
        c += src[i].confidence
      }
      out.push({
        bbox: [x1, y1, x2, y2],
        confidence: c / members.length,
        class_id: targetClassId,
        class_name: params.targetClassName,
      })
    } else {
      for (const i of members) out.push(src[i])
    }
  }

  return out.concat(other).sort((a, b) => b.confidence - a.confidence)
}

File: backend/app/cluster_merge.py
"""
backend/app/cluster_merge.py
=============================
Post-processing pass: collapse dense overlapping detections of one
source class into a single bounding box of a target class.

Use case: the model emits ~50 individual ``ecoli`` boxes packed into a
purple smear; the user wants to see that smear as one big
``ecoli_cluster`` box instead, while loose isolated bacteria stay as
individual ``ecoli`` detections.

Algorithm
---------
1. Pull out every detection whose ``class_id`` matches the source class
   (others are passed through untouched).
2. Inflate each source box by ``margin_frac × max(image_width, image_height)``
   pixels on all sides — a scale-invariant way to define "near".
3. Build connected components via union-find: A and B are connected if
   their inflated boxes intersect. Transitive closure handles A–B–C
   chains automatically.
4. For each component with at least ``min_size`` members, emit ONE box
   equal to the union of the original (non-inflated) corners, labelled
   with the target class. Confidence = mean of member confidences so
   the cluster's score reflects how confident the model was about its
   constituents.
5. Components smaller than ``min_size`` keep their individual
   source-class boxes.

Complexity is O(N²) on source-class boxes. Typical N is ≤ a few hundred
per frame, so this finishes in microseconds.
"""
from typing import List, Tuple

from src.inference import Detection


# ---------------------------------------------------------------------------
# Geometry helpers
# ---------------------------------------------------------------------------

def _inflate(
    bbox: Tuple[float, float, float, float], margin: float,
) -> Tuple[float, float, float, float]:
    x1, y1, x2, y2 = bbox
    return (x1 - margin, y1 - margin, x2 + margin, y2 + margin)


def _intersects(
    a: Tuple[float, float, float, float],
    b: Tuple[float, float, float, float],
) -> bool:
    """Axis-aligned rectangle overlap test (inclusive)."""
    return not (a[2] < b[0] or b[2] < a[0] or a[3] < b[1] or b[3] < a[1])


# ---------------------------------------------------------------------------
# Union-find
# ---------------------------------------------------------------------------

class _UnionFind:
    """Standard union-find with path compression."""

    def __init__(self, n: int):
        self.parent = list(range(n))

    def find(self, i: int) -> int:
        # Iterative path compression — avoids recursion-depth issues on
        # pathological inputs.
        root = i
        while self.parent[root] != root:
            root = self.parent[root]
        while self.parent[i] != root:
            self.parent[i], i = root, self.parent[i]
        return root

    def union(self, i: int, j: int) -> None:
        ri, rj = self.find(i), self.find(j)
        if ri != rj:
            self.parent[ri] = rj


# ---------------------------------------------------------------------------
# Public entry point
# ---------------------------------------------------------------------------

def merge_clusters(
    detections: List[Detection],
    image_size: Tuple[int, int],
    source_class_id: int,
    target_class_id: int,
    target_class_name: str,
    margin_frac: float = 0.01,
    min_size: int = 3,
) -> List[Detection]:
    """Collapse dense ``source_class_id`` detections into clusters.

    Args:
        detections:        raw detector output. Returned unmodified if
                           there's nothing to do.
        image_size:        (width, height) in pixels — needed to make
                           ``margin_frac`` scale-invariant.
        source_class_id:   only boxes of this class are candidates for
                           merging. Others pass through.
        target_class_id:   class_id stamped on emitted cluster boxes.
                           Used by the frontend palette for colour.
        target_class_name: class_name stamped on emitted cluster boxes.
                           Used by the frontend for the label text.
        margin_frac:       neighbour radius as a fraction of the longer
                           image edge. 0.01 = 1% (e.g. ~13 px on a
                           1280×720 frame).
        min_size:          smallest component that becomes a cluster.
                           Components below this stay as individual
                           source-class boxes.

    Returns:
        A new list of ``Detection`` objects, sorted by confidence
        descending so the sidebar reads naturally.
    """
    if not detections or min_size < 2:
        return list(detections)

    # Partition.
    src_dets = [d for d in detections if d.class_id == source_class_id]
    other_dets = [d for d in detections if d.class_id != source_class_id]

    if len(src_dets) < min_size:
        # Not enough source boxes for any cluster — short-circuit.
        return list(detections)

    w, h = image_size
    margin_px = margin_frac * max(w, h)

    inflated = [_inflate(d.bbox, margin_px) for d in src_dets]

    uf = _UnionFind(len(src_dets))
    for i in range(len(src_dets)):
        ai = inflated[i]
        for j in range(i + 1, len(src_dets)):
            if _intersects(ai, inflated[j]):
                uf.union(i, j)

    # Group members by component root.
    groups: dict = {}
    for i in range(len(src_dets)):
        groups.setdefault(uf.find(i), []).append(i)

    out: List[Detection] = []
    for members in groups.values():
        if len(members) >= min_size:
            xs1: List[float] = []
            ys1: List[float] = []
            xs2: List[float] = []
            ys2: List[float] = []
            confs: List[float] = []
            for i in members:
                x1, y1, x2, y2 = src_dets[i].bbox
                xs1.append(x1); ys1.append(y1)
                xs2.append(x2); ys2.append(y2)
                confs.append(src_dets[i].confidence)
            out.append(Detection(
                bbox=(min(xs1), min(ys1), max(xs2), max(ys2)),
                confidence=sum(confs) / len(confs),
                class_id=target_class_id,
                class_name=target_class_name,
            ))
        else:
            for i in members:
                out.append(src_dets[i])

    out.extend(other_dets)
    # Stable, intuitive ordering — highest confidence first.
    out.sort(key=lambda d: -d.confidence)
    return out
